lib/SIL/IR/SILFunctionBuilder.cpp

```diff
@@ -200,29 +201,46 @@ void SILFunctionBuilder::addFunctionAttributes(
   // Only assign replacements when the thing being replaced is function-like and
   // explicitly declared.  
   auto *origDecl = decl->getDynamicallyReplacedDecl();
-  auto *replacedDecl = dyn_cast_or_null<AbstractFunctionDecl>(origDecl);
-  if (!replacedDecl)
-    return;
-
-  // For @objc method replacement we normally use categories to perform the
-  // replacement. Except for methods in generic class where we can't. Instead,
-  // we special case this and use the native swift replacement mechanism.
-  if (decl->isObjC() && !decl->isNativeMethodReplacement()) {
-    F->setObjCReplacement(replacedDecl);
-    return;
-  }
+  if (auto *replacedDecl = dyn_cast_or_null<AbstractFunctionDecl>(origDecl)) {
+    // For @objc method replacement we normally use categories to perform the
+    // replacement. Except for methods in generic class where we can't. Instead,
+    // we special case this and use the native swift replacement mechanism.
+    if (decl->isObjC() && !decl->isNativeMethodReplacement()) {
+      F->setObjCReplacement(replacedDecl);
+      return;
+    }
 
-  if (!constant.canBeDynamicReplacement())
-    return;
+    if (constant.canBeDynamicReplacement()) {
+      SILDeclRef declRef(replacedDecl, constant.kind, false);
+      auto *replacedFunc = getOrCreateDeclaration(replacedDecl, declRef);
 
-  SILDeclRef declRef(replacedDecl, constant.kind, false);
-  auto *replacedFunc = getOrCreateDeclaration(replacedDecl, declRef);
+      assert(replacedFunc->getLoweredFunctionType() ==
+                 F->getLoweredFunctionType() ||
+             replacedFunc->getLoweredFunctionType()->hasOpaqueArchetype());
 
-  assert(replacedFunc->getLoweredFunctionType() ==
-             F->getLoweredFunctionType() ||
-         replacedFunc->getLoweredFunctionType()->hasOpaqueArchetype());
+      F->setDynamicallyReplacedFunction(replacedFunc);
+    }
+  }
 
-  F->setDynamicallyReplacedFunction(replacedFunc);
+  if (constant.isDistributedThunk()) {
+    auto *actor = decl->getDeclContext()->getSelfNominalTypeDecl();
+    if (actor && actor->isDistributedActor()) {
+      auto &C = decl->getASTContext();
+      auto systemTy = getConcreteReplacementForProtocolActorSystemType(decl);
+      assert(systemTy);
+
+      auto decoderTy =
+          getDistributedActorSystemInvocationDecoderType(
+              systemTy->getAnyNominal());
+      assert(decoderTy);
+
+      auto decodeFunc = C.getDecodeNextArgumentOnDistributedInvocationDecoder(
```

Comment:
Nickpick. This is a lot of code inline. Does it make sense to sink the code into a method on actor (ClassDecl?)?

So that we have something like:
```
if (!constant.isDistributedThunk())
  return;
auto *actor = decl->getDeclContext()->getSelfNominalTypeDecl();
if (!actor)
  return;
auto decodeFunc = actor->ifActorGetDecodeNextArgumentOnDistributedInvocationDecode();
if (!decodeFunc)
  return;
auto decodeRef = SILDeclRef(decodeFunc);
...
```

Reply:
Cool, yeah happy to do that -- I always (over)-worry about putting things on ClassDecl but that'll make it much nicer.
I'll refactor a bit, thanks for calling it out 👍

Reply:
Long method name but nicer:

```
    if (auto decodeFunc =
            getAssociatedDistributedInvocationDecoderDecodeNextArgumentFunction(
                decl)) {

      auto decodeRef = SILDeclRef(decodeFunc);
      auto *adHocWitness = getOrCreateDeclaration(decodeFunc, decodeRef);
      F->setReferencedAdHocRequirementWitnessFunction(adHocWitness);
    }

```